File: product-brain/src/product_brain/blocks/hotspot.py
from __future__ import annotations

from collections import Counter, defaultdict
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from typing import Optional

from ..models import TicketRecord


@dataclass
class Cluster:
    files: list[str]
    owners: list[tuple[str, int]]
    avg_loc: float
    avg_duration_days: float
    keywords: list[tuple[str, int]] = field(default_factory=list)
    ticket_ids: list[str] = field(default_factory=list)


def _recency_weight(commit_dt: Optional[datetime], now: datetime) -> float:
    if commit_dt is None:
        return 1.0
    if commit_dt.tzinfo is None:
        commit_dt = commit_dt.replace(tzinfo=timezone.utc)
    age = now - commit_dt
    if age <= timedelta(days=30):
        return 3.0
    if age <= timedelta(days=90):
        return 2.0
    return 1.0


def _drop_mega_files(file_counts: Counter, threshold: float) -> Counter:
    if not file_counts or threshold >= 1.0:
        return file_counts
    sorted_counts = sorted(file_counts.values(), reverse=True)
    cutoff_idx = int(len(sorted_counts) * (1 - threshold))
    if cutoff_idx <= 0:
        return file_counts
    cutoff_value = sorted_counts[cutoff_idx]
    return Counter({f: c for f, c in file_counts.items() if c <= cutoff_value})


def cluster_hotspots(
    records: list[TicketRecord],
    top_k: int = 15,
    cooccur_threshold: float = 0.4,
    mega_file_threshold: float = 0.95,
    now: Optional[datetime] = None,
) -> list[Cluster]:
    """Greedy file clustering by frequency + co-change.

    Mostly deterministic. The LLM-theming step is separate (see render).
    """
    if not records:
        return []
    now = now or datetime.now(timezone.utc)

    file_weights: Counter = Counter()
    file_to_tickets: dict[str, set[str]] = defaultdict(set)
    file_to_authors: dict[str, Counter] = defaultdict(Counter)
    file_loc: dict[str, list[int]] = defaultdict(list)
    file_durations: dict[str, list[float]] = defaultdict(list)

    for rec in records:
        w = _recency_weight(rec.last_commit, now)
        for f in rec.files:
            file_weights[f.path] += w
            file_to_tickets[f.path].add(rec.ticket)
            for a in rec.authors:
                file_to_authors[f.path][a] += 1
            file_loc[f.path].append(f.loc_added + f.loc_removed)
            if rec.duration_days:
                file_durations[f.path].append(rec.duration_days)

    file_weights = _drop_mega_files(file_weights, mega_file_threshold)
    if not file_weights:
        return []

    top_files = [f for f, _ in file_weights.most_common(top_k)]

    cooccur: Counter = Counter()
    for rec in records:
        paths = sorted(set(f.path for f in rec.files) & set(top_files))
        for i, a in enumerate(paths):
            for b in paths[i + 1:]:
                cooccur[(a, b)] += 1
                cooccur[(b, a)] += 1

    clusters: list[Cluster] = []
    used: set[str] = set()

    for seed in top_files:
        if seed in used:
            continue
        cluster_files = [seed]
        used.add(seed)
        seed_count = max(len(file_to_tickets[seed]), 1)
        for f in top_files:
            if f in used:
                continue
            if cooccur[(seed, f)] / seed_count >= cooccur_threshold:
                cluster_files.append(f)
                used.add(f)
        if not cluster_files:
            continue

        owners_counter: Counter = Counter()
        loc_samples: list[int] = []
        duration_samples: list[float] = []
        ticket_ids: set[str] = set()
        for f in cluster_files:
            owners_counter.update(file_to_authors[f])
            loc_samples += file_loc[f]
            duration_samples += file_durations[f]
            ticket_ids |= file_to_tickets[f]

        clusters.append(Cluster(
            files=cluster_files,
            owners=owners_counter.most_common(5),
            avg_loc=(sum(loc_samples) / len(loc_samples)) if loc_samples else 0.0,
            avg_duration_days=(sum(duration_samples) / len(duration_samples)) if duration_samples else 0.0,
            ticket_ids=sorted(ticket_ids),
        ))

    return clusters

File: product-brain/src/blocks/hotspot.ts
import type { TicketRecord } from "../models.js";

export interface Cluster {
  files: string[];
  owners: Array<[string, number]>;
  avgLoc: number;
  avgDurationDays: number;
  keywords: Array<[string, number]>;
  ticketIds: string[];
}

const DAY_MS = 86_400_000;

function recencyWeight(commitDt: Date | undefined, now: Date): number {
  if (!commitDt) return 1.0;
  const ageDays = (now.getTime() - commitDt.getTime()) / DAY_MS;
  if (ageDays <= 30) return 3.0;
  if (ageDays <= 90) return 2.0;
  return 1.0;
}

function dropMegaFiles(weights: Map<string, number>, threshold: number): Map<string, number> {
  if (weights.size === 0 || threshold >= 1.0) return weights;
  const sorted = [...weights.values()].sort((a, b) => b - a);
  const cutoffIdx = Math.floor(sorted.length * (1 - threshold));
  if (cutoffIdx <= 0) return weights;
  const cutoff = sorted[cutoffIdx];
  if (cutoff === undefined) return weights;
  const out = new Map<string, number>();
  for (const [k, v] of weights) {
    if (v <= cutoff) out.set(k, v);
  }
  return out;
}

function increment<K>(m: Map<K, number>, k: K, n = 1): void {
  m.set(k, (m.get(k) ?? 0) + n);
}

function topN<K>(m: Map<K, number>, n: number): Array<[K, number]> {
  return [...m.entries()].sort((a, b) => b[1] - a[1]).slice(0, n);
}

export function clusterHotspots(
  records: TicketRecord[],
  opts: { topK?: number; cooccurThreshold?: number; megaFileThreshold?: number; now?: Date } = {},
): Cluster[] {
  const topK = opts.topK ?? 15;
  const cooccurThreshold = opts.cooccurThreshold ?? 0.4;
  const megaFileThreshold = opts.megaFileThreshold ?? 0.95;
  const now = opts.now ?? new Date();

  if (records.length === 0) return [];

  const weights = new Map<string, number>();
  const fileToTickets = new Map<string, Set<string>>();
  const fileToAuthors = new Map<string, Map<string, number>>();
  const fileLoc = new Map<string, number[]>();
  const fileDurations = new Map<string, number[]>();

  for (const rec of records) {
    const w = recencyWeight(rec.lastCommit, now);
    for (const f of rec.files) {
      increment(weights, f.path, w);
      let t = fileToTickets.get(f.path);
      if (!t) {
        t = new Set();
        fileToTickets.set(f.path, t);
      }
      t.add(rec.ticket);
      let auth = fileToAuthors.get(f.path);
      if (!auth) {
        auth = new Map();
        fileToAuthors.set(f.path, auth);
      }
      for (const a of rec.authors) increment(auth, a);
      let loc = fileLoc.get(f.path);
      if (!loc) {
        loc = [];
        fileLoc.set(f.path, loc);
      }
      loc.push((f.loc_added ?? 0) + (f.loc_removed ?? 0));
      if (rec.durationDays) {
        let d = fileDurations.get(f.path);
        if (!d) {
          d = [];
          fileDurations.set(f.path, d);
        }
        d.push(rec.durationDays);
      }
    }
  }

  const filtered = dropMegaFiles(weights, megaFileThreshold);
  if (filtered.size === 0) return [];
  const topFiles = topN(filtered, topK).map(([f]) => f);
  const topSet = new Set(topFiles);

  const cooccur = new Map<string, number>();
  const cooccurKey = (a: string, b: string): string => `${a}\x00${b}`;
  for (const rec of records) {
    const paths = [...new Set(rec.files.map((f) => f.path))]
      .filter((p) => topSet.has(p))
      .sort();
    for (let i = 0; i < paths.length; i++) {
      const a = paths[i];
      if (!a) continue;
      for (let j = i + 1; j < paths.length; j++) {
        const b = paths[j];
        if (!b) continue;
        increment(cooccur, cooccurKey(a, b));
        increment(cooccur, cooccurKey(b, a));
      }
    }
  }

  const clusters: Cluster[] = [];
  const used = new Set<string>();

  for (const seed of topFiles) {
    if (used.has(seed)) continue;
    const clusterFiles = [seed];
    used.add(seed);
    const seedCount = Math.max(fileToTickets.get(seed)?.size ?? 0, 1);
    for (const f of topFiles) {
      if (used.has(f)) continue;
      const co = cooccur.get(cooccurKey(seed, f)) ?? 0;
      if (co / seedCount >= cooccurThreshold) {
        clusterFiles.push(f);
        used.add(f);
      }
    }
    const ownersCounter = new Map<string, number>();
    const locSamples: number[] = [];
    const durationSamples: number[] = [];
    const ticketIds = new Set<string>();
    for (const f of clusterFiles) {
      const auth = fileToAuthors.get(f);
      if (auth) {
        for (const [k, v] of auth) increment(ownersCounter, k, v);
      }
      locSamples.push(...(fileLoc.get(f) ?? []));
      durationSamples.push(...(fileDurations.get(f) ?? []));
      const t = fileToTickets.get(f);
      if (t) for (const id of t) ticketIds.add(id);
    }
    const avgLoc = locSamples.length
      ? locSamples.reduce((a, b) => a + b, 0) / locSamples.length
      : 0;
    const avgDurationDays = durationSamples.length
      ? durationSamples.reduce((a, b) => a + b, 0) / durationSamples.length
      : 0;
    clusters.push({
      files: clusterFiles,
      owners: topN(ownersCounter, 5),
      avgLoc,
      avgDurationDays,
      keywords: [],
      ticketIds: [...ticketIds].sort(),
    });
  }

  return clusters;
}
